jetty-server/src/test/java/org/eclipse/jetty/server/DetectorConnectionFactoryTest.java

```diff
@@ -0,0 +1,584 @@
+        sslContextFactory.setKeyStorePassword("storepwd");
+        sslContextFactory.setKeyManagerPassword("keypwd");
+        sslContextFactory.start();
+
+        SSLSocketFactory socketFactory = sslContextFactory.getSslContext().getSocketFactory();
+        try (Socket socket = socketFactory.createSocket(_server.getURI().getHost(), _server.getURI().getPort()))
+        {
+            socket.getOutputStream().write(request.getBytes(StandardCharsets.US_ASCII));
+
+            StringBuilder sb = new StringBuilder();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.US_ASCII));
+
+            while (true)
+            {
+                String line = reader.readLine();
+                if (line == null)
+                    break;
+                sb.append(line).append("\n");
+            }
+
+            return sb.length() == 0 ? null : sb.toString();
+        }
+        finally
+        {
+            sslContextFactory.stop();
+        }
+    }
+
+    @Test
+    public void testDetectingSslProxyToHttpNoSslNoProxy() throws Exception
+    {
+        String keystore = MavenTestingUtils.getTestResourceFile("keystore").getAbsolutePath();
+        SslContextFactory sslContextFactory = new SslContextFactory.Server();
+        sslContextFactory.setKeyStorePath(keystore);
+        sslContextFactory.setKeyStorePassword("storepwd");
+        sslContextFactory.setKeyManagerPassword("keypwd");
+
+        ProxyConnectionFactory proxy = new ProxyConnectionFactory(HttpVersion.HTTP_1_1.asString());
+        SslConnectionFactory ssl = new SslConnectionFactory(sslContextFactory, HttpVersion.HTTP_1_1.asString());
+        DetectorConnectionFactory detector = new DetectorConnectionFactory(ssl, proxy);
+        HttpConnectionFactory http = new HttpConnectionFactory();
+
+        start(detector, http);
+
+        String request = "GET /path HTTP/1.1\n" +
+            "Host: server:80\n" +
+            "Connection: close\n" +
+            "\n";
+        String response = getResponse(request);
+
+        assertThat(response, Matchers.containsString("HTTP/1.1 200"));
+    }
+
+    @Test
+    public void testDetectingSslProxyToHttpNoSslWithProxy() throws Exception
+    {
+        String keystore = MavenTestingUtils.getTestResourceFile("keystore").getAbsolutePath();
+        SslContextFactory sslContextFactory = new SslContextFactory.Server();
+        sslContextFactory.setKeyStorePath(keystore);
+        sslContextFactory.setKeyStorePassword("storepwd");
+        sslContextFactory.setKeyManagerPassword("keypwd");
+
+        ProxyConnectionFactory proxy = new ProxyConnectionFactory(HttpVersion.HTTP_1_1.asString());
+        SslConnectionFactory ssl = new SslConnectionFactory(sslContextFactory, HttpVersion.HTTP_1_1.asString());
+        DetectorConnectionFactory detector = new DetectorConnectionFactory(ssl, proxy);
+        HttpConnectionFactory http = new HttpConnectionFactory();
```

Comment:
Move `http` up, and use `http.getProtocol()` rather than `HttpVersion.HTTP_1_1.asString()`.

Reply:
Sure, let me do it.